sros2/sros2/policy/templates/dds/permissions.xsl

```diff
@@ -20,52 +20,75 @@
   </domains>
 </xsl:variable>
 
-<xsl:template match="/policy/profiles">
+<xsl:param name="allow_ros_discovery_topic" select="0"/>
+
+<xsl:template match="/policy/contexts">
   <xsl:variable name="dds">
     <dds xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xsi:noNamespaceSchemaLocation="http://www.omg.org/spec/DDS-SECURITY/20170901/omg_shared_ca_permissions.xsd">
       <permissions>
-        <xsl:for-each select="profile">
-          <xsl:variable name="_ns">
-            <xsl:call-template name="DelimitNamespace">
-              <xsl:with-param name="ns" select="@ns"/>
-            </xsl:call-template>
-          </xsl:variable>
+        <xsl:for-each select="context">
           <xsl:variable name="common_name">
-            <xsl:value-of select="concat($_ns, @node)"/>
+            <xsl:value-of select="@path"/>
           </xsl:variable>
-          <grant name="{$common_name}">
-            <subject_name>CN=<xsl:value-of select="$common_name"/></subject_name>
-            <xsl:copy-of select="$template_validity"/>
-            <xsl:if test="./*[@*='DENY']">
-              <deny_rule>
-                <xsl:copy-of select="$template_domains"/>
-                <xsl:for-each select="./*[@* = 'DENY']">
-                  <xsl:call-template name="TranslatePermissions">
-                    <xsl:with-param name="qualifier" select="'DENY'"/>
-                  </xsl:call-template>
-                </xsl:for-each>
-              </deny_rule>
-            </xsl:if>
-            <xsl:if test="./*[@* = 'ALLOW']">
-              <allow_rule>
-                <xsl:copy-of select="$template_domains"/>
-                <xsl:for-each select="./*[@* = 'ALLOW']">
-                  <xsl:call-template name="TranslatePermissions">
-                    <xsl:with-param name="qualifier" select="'ALLOW'"/>
-                  </xsl:call-template>
-                </xsl:for-each>
-              </allow_rule>
-            </xsl:if>
-            <default>DENY</default>
-          </grant>
+          <xsl:for-each select="profiles">
+            <grant name="{$common_name}">
```

Comment:
Does it actually work to have different grant names for nodes in the same context ?
it seems to me that a single grant would be used for all nodes in the participant.

there seem to be now 3 possible permission files generated:
- permissions files with single grant with all the permissions of the participant (`/single_context`)
- permissions files with single grant for unique node ran standalone
- permissions files with multiple grants, each encompassing a single node (e.g. `permissions/talker_listener/permissions.xml`)
How do you chose to generate the second or the third one ?
In what use case should the one with multiple grants be used?

Reply:
> Does it actually work to have different grant names for nodes in the same context ?

Correct me if I coded this wrong, but here a `<profiles>` tag (plural) must map to a grant, as it is the containing child element that would allow the parent `<context>` tag to have a sequence of unique non-overlapping domain id rules. The `<profile>` tag (singular) is the only element that one-to-one maps to a individual node.

The transform should do all the dds mangling on mass in one dds permission xml tree, then the sros2 api should just find/extract the `grant`(s) with the `name` attribute that match the context path or identity requested.

Could you give/link example inputs/outputs for the three possibilities, so I understand?

Reply:
The code seems to do what I expect it to (aka 1 or 2 based on the policy file provided).
The thing I don't get is 3:
an exampel of such files is this one committed in the PR https://github.com/ros2/sros2/blob/1a6048818002c45e5decd69ded436b41e6e20e73/sros2/test/policies/permissions/talker_listener/permissions.xml
How can they be generated by the CLI?
When are they expected to be used in practice?

My current understanding is that this permission file is generated by `policy_to_permissions.py` for the sole purpose of testing but cannot be generated via the CLI. Is that correct ?

Reply:
In the file you linked, the `<grant>` tags have different `name` attributes, a thus derive from separate `context` elements in a `policy`. The file is what the raw xml tree looks like when returned by the xsl transform.

> but cannot be generated via the CLI. Is that correct ?

Yes, there isn't a way that users could accidentally misuse the CLI to generate a permission for a single context that would include grants from other contexts, as that would just defeat the purpose of separation of privilege/contexts.

Reply:
> Yes, there isn't a way that users could accidentally misuse the CLI to generate a permission.xml file for a single context that would include grants from other contexts, as that would just defeat the purpose of separation of privilege/contexts.

:+1: this is why I was confused to see such files committed.
As these cannot be obtained by users from the CLI that sounds good :+1: thanks for clarifying.